runtime/autoload/spotbugs.vim

```diff
@@ -0,0 +1,184 @@
+          if strlen(line) > 7 && line[0 : 3] == "\xca\xfe\xba\xbe" &&
+                \ ((line[6] == "\n" ? 0 : char2nr(line[6]) % 256) * 256 +
+                    \ char2nr(line[7]) % 256) >= 45
+            echomsg printf('Deleting %s: %d', classname, delete(classname))
+          endif
+        endif
+      endfor
+    finally
+      let &encoding = encoding
+    endtry
+
+    let b:spotbugs_class_files = []
+  endfunction
+endif
+
+function! spotbugs#DefaultPostCompilerAction() abort
+  " Since v7.4.191.
+  make %:S
+endfunction
+
+" Look for "spotbugs#compiler" in "ftplugin/java.vim".
+let s:compiler = exists('spotbugs#compiler') ? spotbugs#compiler : ''
+let s:readable = filereadable($VIMRUNTIME . '/compiler/' . s:compiler . '.vim')
+
+if s:readable && s:compiler ==# 'maven' && executable('mvn')
+
+  function! spotbugs#DefaultPreCompilerAction() abort
+    call spotbugs#DeleteClassFiles()
+    compiler maven
+    make compile
+  endfunction
+
+  function! spotbugs#DefaultPreCompilerTestAction() abort
+    call spotbugs#DeleteClassFiles()
+    compiler maven
+    make test-compile
+  endfunction
+
+  function! spotbugs#DefaultProperties() abort
+    return {
+        \ 'PreCompilerAction':
+            \ function('spotbugs#DefaultPreCompilerAction'),
+        \ 'PreCompilerTestAction':
+            \ function('spotbugs#DefaultPreCompilerTestAction'),
+        \ 'PostCompilerAction':
+            \ function('spotbugs#DefaultPostCompilerAction'),
+        \ 'sourceDirPath':      'src/main/java',
+        \ 'classDirPath':       'target/classes',
+        \ 'testSourceDirPath':  'src/test/java',
+        \ 'testClassDirPath':   'target/test-classes',
+        \ }
+  endfunction
+
+  unlet s:readable s:compiler
+elseif s:readable && s:compiler ==# 'ant' && executable('ant')
+
+  function! spotbugs#DefaultPreCompilerAction() abort
+    call spotbugs#DeleteClassFiles()
+    compiler ant
+    make compile
+  endfunction
+
+  function! spotbugs#DefaultPreCompilerTestAction() abort
+    call spotbugs#DeleteClassFiles()
+    compiler ant
+    make compile-test
+  endfunction
+
+  function! spotbugs#DefaultProperties() abort
+    return {
+        \ 'PreCompilerAction':
+            \ function('spotbugs#DefaultPreCompilerAction'),
+        \ 'PreCompilerTestAction':
+            \ function('spotbugs#DefaultPreCompilerTestAction'),
+        \ 'PostCompilerAction':
+            \ function('spotbugs#DefaultPostCompilerAction'),
+        \ 'sourceDirPath':      'src',
+        \ 'classDirPath':       'build/classes',
+        \ 'testSourceDirPath':  'test',
+        \ 'testClassDirPath':   'build/test/classes',
+        \ }
+  endfunction
+
+  unlet s:readable s:compiler
+elseif s:readable && s:compiler ==# 'javac' && executable('javac')
+
+  function! spotbugs#DefaultPreCompilerAction() abort
+    call spotbugs#DeleteClassFiles()
+    compiler javac
+    make %:S
+  endfunction
```

Comment:
This singling out of a file with `%:S` is bad.  It tells _users_
to go chase type dependencies by hand; let Javac do that.   
We need some way to paste `:args` or something.

Reply:
I'm preparing a fix.

Reply:
Javac compiles dependent classes, but no type dependencies? I'm curious about your fix

Reply:
Grab the posted [example project](https://github.com/vim/vim/pull/16001#issuecomment-2474801801) from above and try compiling  
it as follows from a shell:

```shell
javac --enable-preview --release 23 -Xlint -Xdiags:verbose \
  -cp /tmp/project/bin -d /tmp/project/bin \
  /tmp/project/src/com/acme/EntryPoint.java
```

This is close enough to what `%:S` does.

(If you don't have JDK 23, just replace the `import` novelty,
i.e.

```java
// import module java.base;
import java.util.StringJoiner;

```

and adapt `--release`, `--enable-preview`.)

Nothing surprising.  *You* tell it where dependencies are:

```shell
javac --enable-preview --release 23 -Xlint -Xdiags:verbose \
  -cp /tmp/project/bin -d /tmp/project/bin \
  $(find /tmp/project/src -iname \*.java)
```

Reply:
Here's a patch:

```diff
diff --git a/runtime/autoload/spotbugs.vim b/runtime/autoload/spotbugs.vim
index 47425b47b..28a9ade3d 100644
--- a/runtime/autoload/spotbugs.vim
+++ b/runtime/autoload/spotbugs.vim
@@ -122,6 +122,7 @@ else
 
     let b:spotbugs_class_files = []
   endfunction
+
 endif
 
 function! spotbugs#DefaultPostCompilerAction() abort
@@ -198,7 +199,16 @@ elseif s:readable && s:compiler ==# 'javac' && executable('javac')
   function! spotbugs#DefaultPreCompilerAction() abort
     call spotbugs#DeleteClassFiles()
     compiler javac
-    make %:S
+
+    if get(b:, 'javac_makeprg_params', get(g:, 'javac_makeprg_params', '')) =~ '\s@\S'
+      " Read options and filenames from @options [@sources ...].
+      make
+    else
+      " Let Javac figure out what files to compile.
+      execute 'make ' . join(map(filter(copy(v:argv),
+          \ "v:val =~# '\\.java\\=$'"),
+          \ 'shellescape(v:val)'), ' ')
+    endif
   endfunction
 
   function! spotbugs#DefaultPreCompilerTestAction() abort
```

Now another thing has surfaced after applying it.  If I try  
it for the just mentioned sample project, with `@filelist` or  
without it, two out of two buffers are automatically visited  
with Javac and SpotBugs run for each, but the `Syntax` event  
for the first buffer is not fired (`:au java_spotbugs`).

If I try JDK 21 instead, after making the `import` roll-back,  
the first buffer is only used for Javac and SpotBugs, and no  
`Syntax` event for this buffer is left once you able to move  
around it.

Can it be that `&errorformat` is unaware of `--enable-preview`  
warnings somehow?  Or am I making stuff up?

(Remember to populate `g:javac_makeprg_params`!)